pkg/slo-controller/noderesource/plugins/cpunormalization/plugin.go

```diff
@@ -0,0 +1,299 @@
+limitations under the License.
+*/
+
+package cpunormalization
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	topologyv1alpha1 "github.com/k8stopologyawareschedwg/noderesourcetopology-api/pkg/apis/topology/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/klog/v2"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/koordinator-sh/koordinator/apis/configuration"
+	"github.com/koordinator-sh/koordinator/apis/extension"
+	"github.com/koordinator-sh/koordinator/pkg/slo-controller/noderesource/framework"
+	"github.com/koordinator-sh/koordinator/pkg/util"
+)
+
+const PluginName = "CPUNormalization"
+
+const (
+	defaultRatioStr = "1.00"
+	// in case of unexpected resource amplification
+	// FIXME(saintube): Currently we do not support the scaling factor below 1.0.
+	defaultMinRatio = 1.0
+	defaultMaxRatio = 5.0
+)
+
+var resourcesToAmplify = []corev1.ResourceName{
+	extension.BatchCPU,
+}
+
+var (
+	client     ctrlclient.Client
+	cfgHandler *configHandler
+)
+
+type Plugin struct{}
+
+func (p *Plugin) Name() string {
+	return PluginName
+}
+
+func (p *Plugin) Setup(opt *framework.Option) error {
+	client = opt.Client
+
+	if err := topologyv1alpha1.AddToScheme(opt.Scheme); err != nil {
+		return fmt.Errorf("failed to add scheme for NodeResourceTopology, err: %w", err)
+	}
+	if err := topologyv1alpha1.AddToScheme(clientgoscheme.Scheme); err != nil {
+		return fmt.Errorf("failed to add client go scheme for NodeResourceTopology, err: %w", err)
+	}
+	opt.Builder = opt.Builder.Watches(&source.Kind{Type: &topologyv1alpha1.NodeResourceTopology{}}, &nrtHandler{})
+
+	cfgHandler = newConfigHandler(opt.Client, DefaultCPUNormalizationCfg(), opt.Recorder)
+	opt.Builder = opt.Builder.Watches(&source.Kind{Type: &corev1.ConfigMap{}}, cfgHandler)
+
+	return nil
+}
+
+// NeedSyncMeta checks if the node annotation of cpu normalization ratio to update is different from the current
+func (p *Plugin) NeedSyncMeta(_ *configuration.ColocationStrategy, oldNode, newNode *corev1.Node) (bool, string) {
+	ratioOld, err := extension.GetCPUNormalizationRatio(oldNode)
+	if err != nil {
+		klog.V(4).Infof("failed to get old cpu normalization ratio for node %s, err: %s", oldNode.Name, err)
+		return true, "old ratio parsed error"
+	}
+	ratioNew, err := extension.GetCPUNormalizationRatio(newNode)
+	if err != nil {
+		klog.V(4).Infof("failed to get new cpu normalization ratio for node %s, err: %s", newNode.Name, err)
+		return false, "new ratio parsed error"
+	}
+
+	if ratioNew == -1 && ratioOld == -1 { // no annotation to set
+		return false, "ratios are nil"
+	}
+	if ratioOld == -1 { // annotation to create
+		return true, "old ratio is nil"
+	}
+	if ratioNew == -1 { // annotation to remove
+		return true, "new ratio is nil"
+	}
+	if ratioNew == ratioOld {
+		return false, "ratios unchanged"
```

Comment:
Direct comparison of floating point numbers is actually inappropriate and requires some assumptions.

Reply:
done by differing with an epsilon